DataScience/Analytics/linking/addressLinking.py

```diff
@@ -622,22 +656,69 @@ def parse_input_addresses_to_tokens(self):
+        # if building number is empty and subBuildingName is a only number, add to BuildingStartNumber
+        msk = self.toLinkAddressData['SubBuildingName'].str.contains('\d+', na=False, case=False) & \
+              self.toLinkAddressData['BuildingStartNumber'].isnull()
+        self.toLinkAddressData.loc[msk, 'BuildingStartNumber'] = self.toLinkAddressData.loc[msk, 'SubBuildingName']
+
+        # split flat or apartment number as separate for numerical comparison - compare e.g. SAO number
+        msk = self.toLinkAddressData['SubBuildingName'].str.contains('flat|apartment|unit', na=False, case=False)
         self.toLinkAddressData.loc[msk, 'FlatNumber'] = self.toLinkAddressData.loc[msk, 'SubBuildingName']
         self.toLinkAddressData.loc[msk, 'FlatNumber'] = \
             self.toLinkAddressData.loc[msk].apply(lambda x: x['FlatNumber'].strip().
-                                                  replace('FLAT', '').replace('APARTMENT', ''), axis=1)
+                                                  replace('FLAT', '').replace('APARTMENT', '').replace('UNIT', ''),
+                                                  axis=1)
+
+        # sometimes subBuildingName is e.g. C2 where to number refers to the flat number
+        msk = self.toLinkAddressData['FlatNumber'].str.contains('[A-Z]\d+', na=False, case=False)
+        self.toLinkAddressData.loc[msk, 'FlatNumber'] = \
+            self.toLinkAddressData.loc[msk, 'FlatNumber'].str.replace('[A-Z]', '')
+
+        # deal with addresses that are of type 5/7 4 whatever road...
+        msk = self.toLinkAddressData['SubBuildingName'].str.contains('\d+\/\d+', na=False, case=False) & \
+              self.toLinkAddressData['FlatNumber'].isnull() & ~self.toLinkAddressData['BuildingNumber'].isnull()
+        self.toLinkAddressData.loc[msk, 'FlatNumber'] = \
+            self.toLinkAddressData.loc[msk, 'SubBuildingName'].str.replace('\/\d+', '')
+
+        # some addresses have / as the separator for buildings and flats, when matching against NLP, needs "FLAT"
+        msk = self.toLinkAddressData['SubBuildingName'].str.contains('\d+\/\d+', na=False, case=False)
+        self.toLinkAddressData.loc[msk, 'SubBuildingName'] = 'FLAT ' + \
+                                                             self.toLinkAddressData.loc[msk, 'SubBuildingName']
+
+        # if SubBuildingName contains only numbers, then place also to the flat number field as likely to be flat
+        msk = self.toLinkAddressData['SubBuildingName'].str.isnumeric() & self.toLinkAddressData['FlatNumber'].isnull()
+        msk[msk.isnull()] = False
+        self.toLinkAddressData.loc[msk, 'FlatNumber'] = self.toLinkAddressData.loc[msk, 'SubBuildingName']
+
+        # some addresses, e.g. "5B ELIZABETH AVENUE", have FLAT implicitly even if not spelled -> add "FLAT X"
+        msk = (~self.toLinkAddressData['BuildingSuffix'].isnull()) & \
```

Comment:
Pandas does have .notnull() method (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.notnull.html), so I could use that rather than revert the boolean array. It would probably be more efficient and definitely more readable.